packages/workspace/src/browser/workspace-service.ts

```diff
@@ -57,18 +66,128 @@ export class WorkspaceService implements FrontendApplicationContribution {
 
     @postConstruct()
     protected async init(): Promise<void> {
-        const rootUri = await this.server.getRoot();
-        this._root = await this.toValidRoot(rootUri);
-        if (this._root) {
-            const uri = new URI(this._root.uri);
-            this.updateTitle(uri);
-            this.watcher.watchFileChanges(uri);
+        this._workspaceDescription = {
+            root: undefined,
+            roots: [],
+            workspaceConfigFile: undefined
+        };
+        await this.updateWorkspace();
+        this.updateTitle();
+        this.deferredWorkspaceDescription.resolve(this._workspaceDescription);
+        const configUri = this.getWorkspaceConfigFileUri();
+        if (configUri) {
+            const configUriString = configUri.toString();
+            this.watcher.onFilesChanged(changes => {
+                if (changes.some(change => change.uri.toString() === configUriString)) {
+                    this.updateWorkspace();
+                }
+            });
+        }
+    }
+
+    get root(): Promise<FileStat | undefined> {
+        return this.deferredRoot.promise;
+    }
+    get workspace(): Promise<WorkspaceDescription> {
+        return this.deferredWorkspaceDescription.promise;
+    }
+    get numberOfRoots(): number {
+        return this._workspaceDescription.roots.length;
+    }
+
+    protected readonly onWorkspaceChangeEmitter = new Emitter<WorkspaceDescription>();
+    get onWorkspaceChanged(): Event<WorkspaceDescription> {
+        return this.onWorkspaceChangeEmitter.event;
+    }
+
+    protected async updateWorkspace(): Promise<void> {
+        if (!this._root) {
+            const rootUri = await this.server.getWorkspace();
+            this._root = await this.toValidRoot(rootUri);
         }
+
         this.deferredRoot.resolve(this._root);
+        this._workspaceDescription.root = this._root;
+        await this.preferences.ready;
+        if (this._root) {
+            let roots: string[];
+            let configFileStat = undefined;
+            if (this.preferences['workspace.supportMultiRootWorkspace']) {
```

Comment:
We don't need to check the preference here, as the logic should nicely handle single root ws.

Reply:
As the git commit message explains, i added a new preference 'workspace.supportMultiRootWorkspace' for the user to toggle the multi-root support. It is false by default, as this PR is the minimal patch and there are quite a few packages not updated to accommodate this change. 

I agree that we don't need to check the pref there in future, when the feature is full fledged and 'workspace.supportMultiRootWorkspace' gets removed.

Reply:
Please correct me if I am wrong:
We added a preference "workspace.supportMultiRootWorkspace" to toggle the multi root support. Therefore we either need to check the prefs here, or check the same pref in all the other extensions that access `workspaceService.roots`.